lldb-eval/eval.cc

```diff
@@ -917,15 +924,24 @@ Value Interpreter::EvaluateBinaryBitwiseAssign(BinaryOpKind kind, Value lhs,
   assert(lhs.IsScalar() && "invalid ast: lhs must be an arithmetic type");
   assert(rhs.type().IsBasicType() && "invalid ast: rhs must be a basic type");
 
-  Value ret = CastScalarToBasicType(target_, lhs, rhs.type());
+  Value ret = CastScalarToBasicType(target_, lhs, rhs.type(), error_);
   ret = EvaluateBinaryBitwise(kind, ret, rhs);
-  ret = CastScalarToBasicType(target_, ret, lhs.type());
+  ret = CastScalarToBasicType(target_, ret, lhs.type(), error_);
 
   lhs.Update(ret);
   return lhs;
 }
 
 Value Interpreter::PointerAdd(Value lhs, int64_t offset) {
+  if (lhs.GetUInt64() == 0 && offset != 0) {
+    // This is UB and the compiler would generate a warning:
+    //
+    //  warning: performing pointer arithmetic on a null pointer has undefined
+    //  behaviour if the offset is nonzero [-Wnull-pointer-arithmetic]
+    //
+    error_.SetUbStatus(UbStatus::kNullptrArithmetic);
+  }
```

Comment:
I only encountered this with literal zero / `std::nullptr_t` variables with negative offsets. I think this should still be handled in evaluation, and perhaps we could make this more strict (e.g. check for negative offsets instead of non-zero). If we want to limit this only for literal zero / `nullptr`, `is_literal_zero()` method is accessible in `eval.cc` in `AstNode`, but in this case this we first have to iterate through `CStyleCastNode`s before getting to the literal zero node.

This causes issues, here are some examples:
```
> (char*)0 + short_min
lldb value     : `0x00000000ffff8000`
lldb-eval value: `0xffffffffffff8000`

> (char*)null_ptr + short_min
lldb value     : `0x00000000ffff8000`
lldb-eval value: `0xffffffffffff8000`

// No differences:
> (char*)void_null_ptr + short_min
`0xffffffffffff8000`

> (char*)(1-1) + short_min
`0xffffffffffff8000`
```

This kind of UB happens 5-10 times out of ~100000 expressions (which is way less than other kinds of UBs, but still happens sometimes).